rialto_airflow/funders/linker.py

```diff
@@ -116,35 +114,81 @@ def _find_or_create_funder(snapshot, funder: dict) -> Optional[int]:
-        funder_id = session.execute(
-            insert(Funder)
-            .values(name=name, grid_id=grid_id, federal=federal)
-            .on_conflict_do_update(
-                constraint="funder_grid_id_key", set_=dict(name=name, federal=federal)
-            )
-            .returning(Funder.id)
-        ).scalar_one()
+    funder_id = session.execute(
+        insert(Funder)
+        .values(name=name, grid_id=grid_id, federal=federal)
+        .on_conflict_do_update(
+            constraint="funder_grid_id_key", set_=dict(name=name, federal=federal)
+        )
+        .returning(Funder.id)
+    ).scalar_one()
 
     return funder_id
 
 
-def _lookup_openalex_funder(funder: dict) -> Optional[dict]:
+def _find_or_create_openalex_funder(
+    session: Session, openalex_id: str
+) -> Optional[int]:
+    # if the funder is in the database aleady return it
+    funder = session.execute(
+        select(Funder).where(Funder.openalex_id == openalex_id)
+    ).scalar_one_or_none()
+    if funder:
+        return funder.id
+
+    # otherwise look it up with the api
+    funder = _lookup_openalex_funder(openalex_id)
+    if not funder:
+        logging.warn(f"No funder found in OpenAlex for {openalex_id}")
+        return None
+
+    funder_id = session.execute(
+        insert(Funder)
+        .values(**funder)
+        .on_conflict_do_update(
+            constraint="funder_grid_id_key",
+            # TODO: if deadlocked try removing grid_id from updated values?
+            set_={k: v for k, v in funder.items() if v is not None},
+        )
+        .returning(Funder.id)
+    ).scalar_one()
+
+    return funder_id
+
+
+def _lookup_openalex_funder(openalex_id: str) -> Optional[dict]:
     """
-    Look up funder in OpenAlex to get ROR and then map to GRID ID
+    Look up funder in OpenAlex and return it after trying to map it to GRID ID
+    and looking up whether it appears to be a federal funder.
     """
-    openalex_funder = Funders()[funder.get("funder")]
-    name = openalex_funder.get("display_name")
-    ror = openalex_funder.get("ids", {}).get("ror", None)
-    if ror:
-        grid = convert_ror_to_grid(ror)
-        if grid is None:
-            logging.info(f"missing GRID ID for {funder}")
-            return None
-        # TODO: add ROR ID to funder
-        funder_data = {"id": grid, "name": name}
-
-        return funder_data
-    else:
-        logging.info(f"no ROR ID for {funder}")
+    try:
+        # TODO: get a key so we don't need to do this
+        time.sleep(1)
+        funder = Funders()[openalex_id]
+        logging.info(f"found funder data in openalex for {openalex_id}")
+    except requests.exceptions.HTTPError:
+        logging.warn(f"OpenAlex API returned error for {openalex_id}")
         return None
+
+    # a ror_id is required
+    ror_id = funder.get("ids", {}).get("ror")
+    if ror_id is None:
+        logging.warn(f"no ROR ID for {openalex_id}")
```

Comment:
I think this is a normal situation, especially with small funding orgs, so could be logged at the info level.

Reply:
It might be normal, but currently we are ignoring this Funder information from OpenAlex, which seems not great?

Reply:
Sure, and it will be good to see how many don't have RORs.